plugin/klogr/opt.go

```diff
@@ -0,0 +1,10 @@
+package klogr
+
+type Opt func(l *Logger)
+
+// Debug logging level.
+func Debug() Opt {
+	return func(l *Logger) {
+		l.debug = true
+	}
+}
```

Comment:
I'm going to merge this isn't the New function itself, and support all levels (warn is currently missing, for example)

Reply:
Ahh I noticed that Debug was checked for the extra logging so I added it as a special case, I didn't realise Info was too, I can add it if need?

Reply:
Info's the most common and I'd always recommend it. The problem currently is that Level returns _either_ Error or Debug, and there are a few places within the code where I check the level _before_ calling Log. If the level in klogr is Error, then kgo's Warn and Info will never be called in those cases.

Looking at the logr docs a lot, I think the Level function in this package could be different, and we can actually better support the levels a user has defined.

This is the diff I have locally, wdyt?

```diff
diff --git a/plugin/klogr/klogr.go b/plugin/klogr/klogr.go
index 4c47362..2d9d963 100644
--- a/plugin/klogr/klogr.go
+++ b/plugin/klogr/klogr.go
@@ -1,7 +1,4 @@
-// Package klogr is a wrapper around logr.Logger. The interface of logr.Logger
-// doesn't expose the log level (V-levels in logr), so if you want debug level
-// logging from the kgo.Client, you have to specifically enable it when making
-// a New klogr.Logger.
+// Package klogr is a wrapper around logr.Logger.
 package klogr

 import (
@@ -11,27 +8,55 @@ import (

 var _ kgo.Logger = (*Logger)(nil)

+// Opt applies options to the logger.
+type Opt interface{ apply(*Logger) }
+
+type opt struct{ fn func(*Logger) }
+
+func (o opt) apply(l *Logger) { o.fn(l) }
+
+// LevelToV returns the logr V level that corresponds to the given
+// kgo.LogLevel.
+//
+// By default, the kgo.LogLevel is just doubled -- error is 2, warn is 4, info
+// is 6, and debug is 8.
+func LevelToV(fn func(kgo.LogLevel) int) Opt {
+	return opt{func(l *Logger) { l.levelMap = fn }}
+}
+
 // Logger that wraps the logr.Logger.
 type Logger struct {
-	logr  logr.Logger
-	debug bool
+	logr     logr.Logger
+	levelMap func(kgo.LogLevel) int
 }

 // New returns a new Logger.
 func New(logr logr.Logger, opts ...Opt) *Logger {
-	l := &Logger{logr: logr}
-	for _, o := range opts {
-		o(l)
+	l := &Logger{
+		logr:     logr,
+		levelMap: func(level kgo.LogLevel) int { return int(level) * 2 },
+	}
+	for _, opt := range opts {
+		opt.apply(l)
 	}
 	return l
 }

 // Level returns the log level to log at.
 func (l *Logger) Level() kgo.LogLevel {
-	if l.debug {
-		return kgo.LogLevelDebug
+	// The kgo library really only checks the level ahead of time for debug
+	// logs currently.
+	for _, level := range []kgo.LogLevel{
+		kgo.LogLevelDebug,
+		kgo.LogLevelInfo,
+		kgo.LogLevelWarn,
+		kgo.LogLevelError,
+	} {
+		if l.logr.GetSink().Enabled(l.levelMap(level)) {
+			return level
+		}
 	}
-	return kgo.LogLevelError
+	return kgo.LogLevelNone
 }

 // Log using the underlying logr.Logger. If kgo.LogLevelError is set, keyvals
@@ -39,7 +64,6 @@ func (l *Logger) Level() kgo.LogLevel {
 func (l *Logger) Log(level kgo.LogLevel, msg string, keyvals ...interface{}) {
 	switch level {
 	case kgo.LogLevelNone:
-		return
 	case kgo.LogLevelError:
 		var err error
 		for _, kv := range keyvals {
@@ -50,10 +74,8 @@ func (l *Logger) Log(level kgo.LogLevel, msg string, keyvals ...interface{}) {
 			}
 		}
 		l.logr.Error(err, msg, keyvals...)
-		return
-	}

-	// Pick a V-Level that roughly matches the provided kgo.LogLevel.
-	vLevel := 2 * int(level)
-	l.logr.V(vLevel).Info(msg, keyvals...)
+	default:
+		l.logr.V(l.levelMap(level)).Info(msg, keyvals...)
+	}
 }
diff --git a/plugin/klogr/opt.go b/plugin/klogr/opt.go
deleted file mode 100644
index 9bc15ea..0000000
--- a/plugin/klogr/opt.go
+++ /dev/null
@@ -1,10 +0,0 @@
-package klogr
-
-type Opt func(l *Logger)
-
-// Debug logging level.
-func Debug() Opt {
-	return func(l *Logger) {
-		l.debug = true
-	}
-}
```

Reply:
(I went ahead and force pushed to this branch because it's way easier to look at the diff on GH with before/after of the file)

Reply:
I'll merge this anytime between 2hrs from now and when I go to bed -- please lmk if the current change completely misses the mark for integrating with logr.

Reply:
LGTM